docs/v1/api.md

```diff
@@ -0,0 +1,230 @@
+# Service Catalog API
+
+This document proposes a Kubernetes-native service broker API. It is a formalization of [@krancour](http://github.com/krancour)'s [presentation](https://docs.google.com/presentation/d/1Fm3qG9zQ4R8ZbhtTObmQjO3UZYJ0ruYI4qr5w_gEKN8/edit?usp=sharing) that he gave at the sig-service-catalog meeting on 10/10/2016.
+
+# Motivations
+
+2 different demos were given at [10/3/2016’s sig-service-catalog meeting](https://www.youtube.com/watch?v=Kfar5Uw7CRg), both extremely promising. Each differed in quite a few ways, including how consumers used Kubernetes resources to communicate.
+
+This proposal aims to define a standard API, including a schema for Kubernetes-native resources, to:
+
+- Guide the continuing efforts shown in those demos
+- Improve the user experience (both cluster operator and application developer)
+- Attempt to extend the Kubernetes API instead of relying on comparatively clunky Third Party Resource types (long-term goal)
+- Ensure a consistent user experience regardless of implementation, without undermining competition or choice
+
+# Inspiration
+
+Broadly speaking, we're proposing a system that can list services and allow consumers to provision, bind, unbind and deprovision them.
+
+We'll draw inspiration from the [Persistent Volumes](http://kubernetes.io/docs/user-guide/persistent-volumes/) API and the [Ingress Controllers](http://kubernetes.io/docs/user-guide/ingress/) system. The remainder of this section explains these inspirations in further detail.
+
+## Persistent Volumes
+
+First, this problem space is symmetric to that of persistent volumes. We propose to model the data structures herein after those in persistent volumes primarily because we believe that consistency is a virtue. More importantly, the community benefits when similar problems have similar solutions.
+
+Specifically, we believe that the following existing types provide a good model for a service catalog API:
+
+- `StorageClass`
+- `PersistentVolume`
+- `PersistentVolumeClaim`
+
+## Ingress Controllers
+
+Today, operators choose their own implementation of an ingress controller. Kubernetes gives the operator the ability to specify any ingress controller, and the complete freedom to implement it themselves.
```

Comment:
I agree that freedom to implement one's own controller is a good thing, but I _do_ think that we should provide a controller that handles the CF Service Broker 2.0 API.  An ecosystem already exists around this, and we should be able to agree on the mechanics of how they should be used in Kubernetes.

So, just to be clear, I guess what I am really proposing here is that we make it possible not just to implement your own controller but to run multiple controllers that cooperate to handle the _right_ resources.  Then you can run whatever combination you want in the same cluster.  In this regard the problem space is arguably a little different from ingress in that most operators will only want to deal with a single type of ingress controller, but I can easily see people wanting to run multiple service controllers to get different things.

For example, say you have two types of behaviors you want from service catalog.  You want to consume CF services, and you want to use another controller that can do whitebox services, both at the same time.  It should be possible to run those two controllers and ensure that they both do the right thing.

You could implement this as an annotation on a `ServiceBroker` type that would ripple down to the `ServiceClass` instances for that broker.

I don't think we need to formalize this now or anything, but I want to plant the seed.

cc @duglin @slack